lib/Slave.php

```diff
@@ -0,0 +1,244 @@
+								IClientService $clientService,
+								GlobalSiteSelector $gss,
+								ILogger $logger
+	) {
+		$this->accountManager = $accountManager;
+		$this->userManager = $userManager;
+		$this->clientService = $clientService;
+		$this->logger = $logger;
+		$this->lookupServer = $gss->getLookupServerUrl();
+		$this->operationMode = $gss->getMode();
+		$this->authKey = $gss->getJwtKey();
+		$this->lookupServer = rtrim($this->lookupServer, '/');
+		$this->lookupServer .= '/gs/users';
+	}
+
+	public function createUser(array $params) {
+		if ($this->checkConfiguration() === false)  {
+			return;
+		}
+
+		$uid = $params['uid'];
+		$user = $this->userManager->get($uid);
+		$userData = [];
+		if ($user !== null) {
+			$userData[$user->getCloudId()] = $this->getAccountData($user);
+			$this->addUsers($userData);
+		}
+	}
+
+	/**
+	 * the server indicated that the admin want to remove a user, remember the
+	 * federated cloud id so that we can remove the user from the lookup server
+	 * once they were deleted
+	 *
+	 * @param array $params
+	 */
+	public function preDeleteUser(array $params) {
+		$uid = $params['uid'];
+		$user = $this->userManager->get($uid);
+		if ($user !== null) {
+			self::$toRemove[$uid] = $user->getCloudId();
+		}
+	}
+
+	/**
+	 * remove user from lookup server
+	 *
+	 * @param array $params
+	 */
+	public function deleteUser(array $params) {
+		if ($this->checkConfiguration() === false)  {
+			return;
+		}
+
+		$uid = $params['uid'];
+		if (isset(self::$toRemove[$uid])) {
+			$this->removeUsers([self::$toRemove[$uid]]);
+			unset(self::$toRemove[$uid]);
+		}
+	}
+
+	/**
+	 * update the lookup server with all known users on this instance. This
+	 * is triggered by a cronjob
+	 */
+	public function batchUpdate() {
+		if ($this->checkConfiguration() === false)  {
+			return;
+		}
+
+		$backends = $this->userManager->getBackends();
+		foreach ($backends as $backend) {
+			$limit = 200;
+			$offset = 0;
+			$usersData = [];
+			do {
+				$users = $backend->getUsers('', $limit, $offset);
+				foreach ($users as $uid) {
+					$user = $this->userManager->get($uid);
+					if ($user !== null) {
+						$usersData[$user->getCloudId()] = $this->getAccountData($user);
+					}
+				}
+				$offset += $limit;
+				$this->addUsers($usersData);
+			} while (count($users) >= $limit);
+		}
+	}
+
+	/**
+	 * get user data from account manager
+	 *
+	 * @param IUser $user
+	 * @return array
+	 */
+	protected function getAccountData(IUser $user) {
+		$rawData = $this->accountManager->getUser($user);
+		$data = [];
+		foreach ($rawData as $key => $value) {
+			if ($key === 'displayname') {
+				$data['name'] = $value['value'];
+			} else {
+				$data[$key] = $value['value'];
+			}
+		}
+		unset($data['avatar']);
+		return $data;
+	}
+
+	/**
+	 * send users to the lookup server
+	 *
+	 * @param array $users
+	 */
+	protected function addUsers(array $users) {
+		$dataBatch = ['authKey' => $this->authKey, 'users' => $users];
```

Comment:
I just realised this: That is not good. The JWT key should never send over the wire. The idea is to encrypt something with this key and then send this over. Also auth stuff should be send in headers and not in the body, otherwise GET requests can't use this.